flink-staging/flink-gelly/src/main/java/org/apache/flink/graph/gsa/GatherSumApplyIteration.java

```diff
@@ -0,0 +1,374 @@
+			GatherSumApplyIteration<K, VV, EV, M> withEdges(DataSet<Edge<K, EV>> edges,
+			GatherFunction<VV, EV, M> gather, SumFunction<VV, EV, M> sum, ApplyFunction<VV, EV, M> apply,
+			int maximumNumberOfIterations) {
+		return new GatherSumApplyIteration<K, VV, EV, M>(gather, sum, apply, edges, maximumNumberOfIterations);
+	}
+
+	// --------------------------------------------------------------------------------------------
+	//  Triplet Utils
+	// --------------------------------------------------------------------------------------------
+
+	private static final class PairJoinFunction<K extends Comparable<K> & Serializable, VV extends Serializable,
+			EV extends Serializable> implements FlatJoinFunction<Vertex<K, VV>, Edge<K, EV>,
+			Tuple3<K, Vertex<K, VV>, Edge<K, EV>>> {
+
+		@Override
+		public void join(Vertex<K, VV> vertex, Edge<K, EV> edge,
+				Collector<Tuple3<K, Vertex<K, VV>, Edge<K, EV>>> collector) throws Exception {
+			collector.collect(new Tuple3<K, Vertex<K, VV>, Edge<K, EV>>(edge.getTarget(), vertex, edge));
+		}
+	}
+
+	private static final class TripletJoinFunction<K extends Comparable<K> & Serializable, VV extends Serializable,
+			EV extends Serializable> implements  FlatJoinFunction<Tuple3<K, Vertex<K, VV>, Edge<K, EV>>, Vertex<K, VV>,
+			Tuple3<Vertex<K, VV>, Edge<K, EV>, Vertex<K, VV>>> {
+		@Override
+		public void join(Tuple3<K, Vertex<K, VV>, Edge<K, EV>> vertexEdge, Vertex<K, VV> vertex,
+				Collector<Tuple3<Vertex<K, VV>, Edge<K, EV>, Vertex<K, VV>>> out) throws Exception {
+			out.collect(new Tuple3<Vertex<K, VV>, Edge<K, EV>, Vertex<K, VV>>(
+					vertexEdge.f1, vertexEdge.f2, vertex
+			));
+		}
+	}
+
+	// --------------------------------------------------------------------------------------------
+	//  Wrapping UDFs
+	// --------------------------------------------------------------------------------------------
+
+	private static final class GatherUdf<K extends Comparable<K> & Serializable, VV extends Serializable,
+			EV extends Serializable, M> extends RichMapFunction<Tuple3<Vertex<K, VV>, Edge<K, EV>, Vertex<K, VV>>,
+			Tuple2<K, M>> implements ResultTypeQueryable<Tuple2<K, M>> {
+
+		private final GatherFunction<VV, EV, M> gatherFunction;
+		private transient TypeInformation<Tuple2<K, M>> resultType;
+
+		private GatherUdf(GatherFunction<VV, EV, M> gatherFunction, TypeInformation<Tuple2<K, M>> resultType) {
+			this.gatherFunction = gatherFunction;
+			this.resultType = resultType;
+		}
+
+		@Override
+		public Tuple2<K, M> map(Tuple3<Vertex<K, VV>, Edge<K, EV>, Vertex<K, VV>> triplet) throws Exception {
+			Triplet<VV, EV> userTriplet = new Triplet<VV, EV>(triplet.f0.getValue(),
+					triplet.f1.getValue(), triplet.f2.getValue());
+
+			K key = triplet.f2.getId();
+			M result = this.gatherFunction.gather(userTriplet);
+			return new Tuple2<K, M>(key, result);
+		}
+
+		@Override
+		public void open(Configuration parameters) throws Exception {
```

Comment:
@vasia,

Perhaps we should check if the superstep number bug in vertexCentricIteration is reproducible, because from the looks of this, it should be...